### PR TITLE
Lenstransform

### PR DESCRIPTION
**Pull request by jw3126**
Implementing @tkf gist https://github.com/jw3126/Setfield.jl/pull/92#issuecomment-540434449

### CODE REVIEW

File: src/sugar.jl [hunk 1]
```diff
@@ -133,7 +133,7 @@ struct _UpdateOp{OP,V}
 end
 (u::_UpdateOp)(x) = u.op(x, u.val)
 
-function atset_impl(ex::Expr; overwrite::Bool)
+function atset_impl(ex::Expr; overwrite::Bool=false, lenstransform=identity)
```
Comment (jw3126) [inline src/sugar.jl]: Let's rename these functions. How about `lensmacro`/`setmacro`? Also a docstring should be added.

File: test/test_custom_macros.jl [hunk 2]
```diff
@@ -0,0 +1,78 @@
+module CustomMacros
```
Comment (jw3126) [inline test/test_custom_macros.jl]: I think this could be a `Literate` page in the docs.

File: src/sugar.jl [hunk 3]
```diff
@@ -142,14 +142,15 @@ function atset_impl(ex::Expr; overwrite::Bool)
...
     ret = if ex.head == :(=)
         quote
-            lens = $lens
+            lens = ($lenstransform)($lens)
             $dst = set($obj, lens, $val)
         end
     else
         op = get_update_op(ex.head)
-        f = :(_UpdateOp($op,$val))
+        f = :($_UpdateOp($op,$val))
```
Comment (tkf) [inline src/sugar.jl]: I think changes like this need to happen in `parse_obj_lenses` as well. For example, `lens = :($IndexLens($index))`.

File: src/sugar.jl [hunk 4]
```diff
@@ -188,12 +189,16 @@ julia> set(t, (@lens _[1]), "1")
 
 """
 macro lens(ex)
+    atlens_impl(ex)
```
Comment (tkf) [inline src/sugar.jl]: I guess we don't need to expose `atlens_impl` as people can just do

```julia
macro mylens(ex)
    :(mylenstransform(@lens $ex))
end
```

? Or maybe with some more `$` and `esc`. :)

File: test/test_custom_macros.jl (hunk 2 above)
Comment (tkf) [reply test/test_custom_macros.jl]: That's a great idea!

We also should define a macro in an isolated sub-module and import it into the namespace we use for testing. This is for catching some possible namespacing issues.

File: src/sugar.jl (hunk 1 above)
Comment (tkf) [reply src/sugar.jl]: `setmacro` sounds good.

File: src/sugar.jl (hunk 4 above)
Comment (jw3126) [reply src/sugar.jl]: It is nice, that this is a one liner. On the other hand the fact that you are asking about `$` and `esc` shows that it is somewhat non trivial to get isolation right. So I left `lensmacro` alive.

File: examples/custom_macros.jl [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+@test o.b == 20
+
+l = Lens!(@lens _.foo[1])
+o = (foo=[1,2,3], bar=:bar)
+set(o, l, 100)
+@test o == (foo=[100,2,3], bar=:bar)
+
+# Now we can implement the syntax macros
+
+using Setfield: setmacro, lensmacro
```
Comment (tkf) [inline examples/custom_macros.jl]:
```suggestion
module Clone
using Setfield: setmacro, lensmacro
```

File: examples/custom_macros.jl [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+macro myset!(ex)
+    setmacro(Lens!, ex)
+end
+
+macro mylens!(ex)
+    lensmacro(Lens!, ex)
+end
+
```
Comment (tkf) [inline examples/custom_macros.jl]:
```suggestion

end#module Clone

using .Clone: Clone
```

**Pull request merged**